File: api-gateway/routes/notifications.py
"""API Gateway — deadline notifications.

Filing deadlines were calculated and then only shown to whoever happened to
open the dashboard. This turns them into something that arrives.

Two kinds:

  deadline_digest      to the person who owns the work: what of theirs is due
                       soon or already past.
  overdue_escalation   to managers: anything overdue that nobody owns, because
                       an unassigned overdue denial has no one to chase it.

Generation is idempotent by (user, kind, day). The digest runs from cron and
the API runs four workers, so "once a day" has to be a database constraint
rather than an assumption about how often this is called.
"""

import json
import logging
from uuid import UUID
import os
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query, Request
from pydantic import BaseModel

from api_gateway.services.db import get_connection
from api_gateway.services.access import MANAGER_UP, principal
from api_gateway.services.audit import record as audit_record

logger = logging.getLogger("api_gateway.notifications")

router = APIRouter()

# How far ahead a digest looks. Matches the dashboard's priority window, so the
# two cannot tell a user different things about the same denial.
DIGEST_HORIZON_DAYS = int(os.environ.get("DEADLINE_DIGEST_DAYS", "14"))


def _me(http_request: Request) -> str:
    who = getattr(http_request.state, "principal", None) or principal(http_request)
    if who.kind != "user" or not who.user_id:
        raise HTTPException(status_code=401, detail="Not authenticated")
    return who.user_id


@router.get("/notifications", response_model=list[dict])
async def list_notifications(
    http_request: Request,
    unread_only: bool = Query(False),
    limit: int = Query(50, ge=1, le=200),
):
    """Your notifications. Nobody can read anyone else's."""
    user_id = _me(http_request)
    async with get_connection() as conn:
        rows = await conn.fetch(
            f"""
            SELECT id, kind, for_date, title, body, payload, read_at, created_at
              FROM notifications
             WHERE user_id = $1::uuid
               {"AND read_at IS NULL" if unread_only else ""}
             ORDER BY created_at DESC
             LIMIT $2
            """,
            user_id, limit,
        )
    out = []
    for r in rows:
        d = dict(r)
        d["id"] = str(d["id"])
        d["for_date"] = d["for_date"].isoformat() if d["for_date"] else None
        d["created_at"] = d["created_at"].isoformat() if d["created_at"] else None
        d["read_at"] = d["read_at"].isoformat() if d["read_at"] else None
        if isinstance(d.get("payload"), str):
            try:
                d["payload"] = json.loads(d["payload"])
            except ValueError:
                pass
        out.append(d)
    return out


@router.post("/notifications/{notification_id}/read", response_model=dict)
async def mark_read(notification_id: UUID, http_request: Request):
    """Mark one as read. Scoped to the owner, so an id is not enough."""
    user_id = _me(http_request)
    async with get_connection() as conn:
        row = await conn.fetchrow(
            """UPDATE notifications SET read_at = NOW()
                WHERE id = $1::uuid AND user_id = $2::uuid AND read_at IS NULL
            RETURNING id""",
            notification_id, user_id,
        )
    return {"status": "read" if row else "not_found_or_already_read"}


@router.post("/notifications/read-all", response_model=dict)
async def mark_all_read(http_request: Request):
    user_id = _me(http_request)
    async with get_connection() as conn:
        result = await conn.execute(
            "UPDATE notifications SET read_at = NOW() WHERE user_id = $1::uuid AND read_at IS NULL",
            user_id,
        )
    return {"status": "ok", "marked": int(result.split()[-1])}


@router.post("/notifications/generate-digests", response_model=dict)
async def generate_digests(http_request: Request):
    """Build today's deadline digests.

    Called from cron (see scripts/send_deadline_digests.sh). Safe to call more
    than once: the unique index means a second run updates nothing rather than
    sending everyone a duplicate.
    """
    who = getattr(http_request.state, "principal", None) or principal(http_request)
    # A service credential or an admin. Not something a specialist triggers,
    # since it writes to everyone's notifications.
    if not (who.kind == "service" or who.role == "admin"):
        raise HTTPException(status_code=403, detail="Admin or service access required")

    created, escalated = 0, 0
    async with get_connection() as conn:
        # ── per-owner digests ──
        owners = await conn.fetch(
            f"""
            SELECT aq.assigned_user_id AS user_id,
                   COUNT(*) FILTER (WHERE d.appeal_deadline < CURRENT_DATE) AS overdue,
                   COUNT(*) FILTER (WHERE d.appeal_deadline >= CURRENT_DATE) AS upcoming,
                   MIN(d.appeal_deadline) AS soonest,
                   SUM(d.charge_amount) AS amount,
                   json_agg(json_build_object(
                       'claim_number', c.claim_number,
                       'deadline', d.appeal_deadline,
                       'amount', d.charge_amount
                   ) ORDER BY d.appeal_deadline) AS items
              FROM appeals_queue aq
              JOIN denials d ON d.id = aq.denial_id
              JOIN claims c ON c.id = d.claim_id
             WHERE aq.assigned_user_id IS NOT NULL
               AND (aq.outcome_status IS NULL
                    OR aq.outcome_status NOT IN ('approved','overruled','resolved','denied_again','cancelled'))
               AND d.appeal_deadline IS NOT NULL
               AND d.appeal_deadline <= CURRENT_DATE + INTERVAL '{DIGEST_HORIZON_DAYS} days'
             GROUP BY aq.assigned_user_id
            """
        )
        for row in owners:
            overdue, upcoming = row["overdue"], row["upcoming"]
            title = (
                f"{overdue} overdue and {upcoming} due within {DIGEST_HORIZON_DAYS} days"
                if overdue else
                f"{upcoming} filing deadline(s) within {DIGEST_HORIZON_DAYS} days"
            )
            body = (
                f"Soonest is {row['soonest']}. "
                f"{float(row['amount'] or 0):,.2f} in denied charges is affected."
            )
            result = await conn.execute(
                """
                INSERT INTO notifications (user_id, kind, title, body, payload)
                VALUES ($1, 'deadline_digest', $2, $3, $4::jsonb)
                ON CONFLICT (user_id, kind, for_date) DO NOTHING
                """,
                row["user_id"], title, body,
                json.dumps({"overdue": overdue, "upcoming": upcoming,
                            "items": json.loads(row["items"]) if isinstance(row["items"], str) else row["items"]},
                           default=str),
            )
            created += 1 if result.endswith(" 1") else 0

        # ── unowned overdue work goes to managers ──
        orphan = await conn.fetchrow(
            """
            SELECT COUNT(*) AS n, SUM(d.charge_amount) AS amount, MIN(d.appeal_deadline) AS oldest
              FROM denials d
              LEFT JOIN appeals_queue aq ON aq.denial_id = d.id
                   AND (aq.outcome_status IS NULL
                        OR aq.outcome_status NOT IN ('approved','overruled','resolved','denied_again','cancelled'))
             WHERE d.appeal_deadline IS NOT NULL
               AND d.appeal_deadline < CURRENT_DATE
               AND d.status IN ('open', 'analyzed')
               AND (aq.id IS NULL OR aq.assigned_user_id IS NULL)
            """
        )
        if orphan and orphan["n"]:
            managers = await conn.fetch(
                "SELECT id FROM users WHERE is_active AND role = ANY($1::text[])",
                list(MANAGER_UP),
            )
            for m in managers:
                result = await conn.execute(
                    """
                    INSERT INTO notifications (user_id, kind, title, body, payload)
                    VALUES ($1, 'overdue_escalation', $2, $3, $4::jsonb)
                    ON CONFLICT (user_id, kind, for_date) DO NOTHING
                    """,
                    m["id"],
                    f"{orphan['n']} overdue denial(s) with nobody assigned",
                    (f"Oldest deadline {orphan['oldest']}. "
                     f"{float(orphan['amount'] or 0):,.2f} in denied charges is unowned."),
                    json.dumps({"count": orphan["n"], "oldest": str(orphan["oldest"])}),
                )
                escalated += 1 if result.endswith(" 1") else 0

    await audit_record(
        action="deadline_digests_generated",
        resource_type="system",
        details={"digests": created, "escalations": escalated,
                 "horizon_days": DIGEST_HORIZON_DAYS},
    )
    return {"status": "ok", "digests_created": created, "escalations_created": escalated,
            "horizon_days": DIGEST_HORIZON_DAYS}

File: crates/api-gateway/src/routes/notifications.rs
//! Deadline notifications.
//!
//! Ported from `api-gateway/routes/notifications.py`. Two kinds: a
//! `deadline_digest` to the person who owns the work, and an
//! `overdue_escalation` to managers for overdue work nobody owns. Generation
//! is idempotent by `(organization, user, kind, day)` - the unique index, not an assumption
//! about how often the cron calls this, is what makes "once a day" hold.

use axum::extract::{Path, Query, State};
use axum::routing::{get, post};
use axum::{Extension, Json, Router};
use chrono::{DateTime, NaiveDate, Utc};
use denial_auth::rbac::{Principal, PrincipalKind};
use denial_common::error::AppError;
use serde::Deserialize;
use sqlx::Row;
use uuid::Uuid;

use crate::state::AppState;

/// system/revenue-cycle administrators - the roles an unowned overdue
/// denial is escalated to.
const MANAGER_UP: &[&str] = &["revenue_cycle_manager", "system_admin"];

#[derive(Deserialize)]
pub struct ListQuery {
    #[serde(default)]
    pub unread_only: bool,
    #[serde(default = "default_limit")]
    pub limit: i64,
}

fn default_limit() -> i64 {
    50
}

/// The caller's own user id, or 401. Nobody reads anyone else's notifications.
fn me(principal: &Principal) -> Result<Uuid, AppError> {
    if principal.kind != PrincipalKind::User {
        return Err(AppError::Unauthorized);
    }
    principal
        .user_id
        .as_deref()
        .and_then(|s| Uuid::parse_str(s).ok())
        .ok_or(AppError::Unauthorized)
}

fn organization_id(principal: &Principal) -> Result<Uuid, AppError> {
    principal
        .organization_id
        .as_deref()
        .and_then(|s| Uuid::parse_str(s).ok())
        .ok_or(AppError::Forbidden)
}

/// `12345.6` -> `"12,345.60"`, matching Python's `{:,.2f}`.
fn money(v: f64) -> String {
    let neg = v.is_sign_negative();
    let cents = format!("{:.2}", v.abs());
    let (int_part, frac) = cents.split_once('.').unwrap_or((cents.as_str(), "00"));
    let bytes = int_part.as_bytes();
    let mut grouped = String::new();
    for (i, b) in bytes.iter().enumerate() {
        if i > 0 && (bytes.len() - i) % 3 == 0 {
            grouped.push(',');
        }
        grouped.push(*b as char);
    }
    format!("{}{grouped}.{frac}", if neg { "-" } else { "" })
}

pub async fn list_notifications(
    State(state): State<AppState>,
    Extension(principal): Extension<Principal>,
    Query(params): Query<ListQuery>,
) -> Result<Json<Vec<serde_json::Value>>, AppError> {
    let user_id = me(&principal)?;
    let organization_id = organization_id(&principal)?;
    let limit = params.limit.clamp(1, 200);

    let sql = format!(
        "SELECT id, kind, for_date, title, body, payload, read_at, created_at \
           FROM notifications \
          WHERE organization_id = $1 AND user_id = $2 {} \
           ORDER BY created_at DESC \
           LIMIT $3",
        if params.unread_only {
            "AND read_at IS NULL"
        } else {
            ""
        }
    );

    let rows = sqlx::query(&sql)
        .bind(organization_id)
        .bind(user_id)
        .bind(limit)
        .fetch_all(&state.pool)
        .await
        .map_err(AppError::Db)?;

    let out = rows
        .iter()
        .map(|r| {
            let id: Option<Uuid> = r.try_get("id").ok().flatten();
            let for_date: Option<NaiveDate> = r.try_get("for_date").ok().flatten();
            let created_at: Option<DateTime<Utc>> = r.try_get("created_at").ok().flatten();
            let read_at: Option<DateTime<Utc>> = r.try_get("read_at").ok().flatten();
            serde_json::json!({
                "id": id.map(|u| u.to_string()),
                "kind": r.try_get::<Option<String>, _>("kind").ok().flatten(),
                "for_date": for_date.map(|d| d.to_string()),
                "title": r.try_get::<Option<String>, _>("title").ok().flatten(),
                "body": r.try_get::<Option<String>, _>("body").ok().flatten(),
                "payload": r
                    .try_get::<Option<serde_json::Value>, _>("payload")
                    .ok()
                    .flatten()
                    .unwrap_or(serde_json::Value::Null),
                "read_at": read_at.map(|t| t.to_rfc3339()),
                "created_at": created_at.map(|t| t.to_rfc3339()),
            })
        })
        .collect();

    Ok(Json(out))
}

pub async fn mark_read(
    State(state): State<AppState>,
    Extension(principal): Extension<Principal>,
    Path(notification_id): Path<Uuid>,
) -> Result<Json<serde_json::Value>, AppError> {
    let user_id = me(&principal)?;
    let organization_id = organization_id(&principal)?;
    let row = sqlx::query(
        "UPDATE notifications SET read_at = NOW() \
          WHERE id = $1 AND organization_id = $2 AND user_id = $3 AND read_at IS NULL \
        RETURNING id",
    )
    .bind(notification_id)
    .bind(organization_id)
    .bind(user_id)
    .fetch_optional(&state.pool)
    .await
    .map_err(AppError::Db)?;

    Ok(Json(serde_json::json!({
        "status": if row.is_some() { "read" } else { "not_found_or_already_read" }
    })))
}

pub async fn mark_all_read(
    State(state): State<AppState>,
    Extension(principal): Extension<Principal>,
) -> Result<Json<serde_json::Value>, AppError> {
    let user_id = me(&principal)?;
    let organization_id = organization_id(&principal)?;
    let result = sqlx::query(
        "UPDATE notifications SET read_at = NOW() \
          WHERE organization_id = $1 AND user_id = $2 AND read_at IS NULL",
    )
    .bind(organization_id)
    .bind(user_id)
    .execute(&state.pool)
    .await
    .map_err(AppError::Db)?;

    Ok(Json(serde_json::json!({
        "status": "ok",
        "marked": result.rows_affected(),
    })))
}

pub async fn generate_digests(
    State(state): State<AppState>,
    Extension(principal): Extension<Principal>,
) -> Result<Json<serde_json::Value>, AppError> {
    // A service credential or an admin - not something a specialist triggers,
    // since it writes to everyone's notifications.
    let allowed = principal.kind == PrincipalKind::Service
        || matches!(
            principal.role.as_deref(),
            Some("system_admin" | "security_admin")
        );
    if !allowed {
        return Err(AppError::Forbidden);
    }

    let horizon = state.digest_horizon_days as i64;
    let mut created: i64 = 0;
    let mut escalated: i64 = 0;

    // ── per-owner digests ──
    let owners = sqlx::query(&format!(
        "SELECT c.organization_id, aq.assigned_user_id AS user_id, \
                COUNT(*) FILTER (WHERE d.appeal_deadline < CURRENT_DATE)  AS overdue, \
                COUNT(*) FILTER (WHERE d.appeal_deadline >= CURRENT_DATE) AS upcoming, \
                MIN(d.appeal_deadline) AS soonest, \
                SUM(d.charge_amount)::float8 AS amount, \
                json_agg(json_build_object( \
                    'claim_number', c.claim_number, \
                    'deadline', d.appeal_deadline, \
                    'amount', d.charge_amount \
                ) ORDER BY d.appeal_deadline) AS items \
           FROM appeals_queue aq \
           JOIN denials d ON d.id = aq.denial_id \
           JOIN claims c ON c.id = d.claim_id \
          WHERE aq.assigned_user_id IS NOT NULL \
            AND (aq.outcome_status IS NULL \
                 OR aq.outcome_status NOT IN ('approved','overruled','resolved','denied_again','cancelled')) \
            AND d.appeal_deadline IS NOT NULL \
            AND d.appeal_deadline <= CURRENT_DATE + INTERVAL '{horizon} days' \
           GROUP BY c.organization_id, aq.assigned_user_id"
    ))
    .fetch_all(&state.pool)
    .await
    .map_err(AppError::Db)?;

    for row in &owners {
        let org_id: Uuid = row
            .try_get("organization_id")
            .map_err(|e| AppError::Internal(e.to_string()))?;
        let owner_id: Uuid = row
            .try_get("user_id")
            .map_err(|e| AppError::Internal(e.to_string()))?;
        let overdue: i64 = row.try_get("overdue").unwrap_or(0);
        let upcoming: i64 = row.try_get("upcoming").unwrap_or(0);
        let soonest: Option<NaiveDate> = row.try_get("soonest").ok().flatten();
        let amount: f64 = row
            .try_get::<Option<f64>, _>("amount")
            .ok()
            .flatten()
            .unwrap_or(0.0);
        let items: serde_json::Value = row
            .try_get::<Option<serde_json::Value>, _>("items")
            .ok()
            .flatten()
            .unwrap_or_else(|| serde_json::json!([]));

        let title = if overdue > 0 {
            format!("{overdue} overdue and {upcoming} due within {horizon} days")
        } else {
            format!("{upcoming} filing deadline(s) within {horizon} days")
        };
        let body = format!(
            "Soonest is {}. {} in denied charges is affected.",
            soonest.map(|d| d.to_string()).unwrap_or_default(),
            money(amount),
        );
        let payload = serde_json::json!({
            "overdue": overdue,
            "upcoming": upcoming,
            "items": items,
        });

        let result = sqlx::query(
            "INSERT INTO notifications (organization_id, user_id, kind, title, body, payload) \
             VALUES ($1, $2, 'deadline_digest', $3, $4, $5::jsonb) \
             ON CONFLICT (organization_id, user_id, kind, for_date) DO NOTHING",
        )
        .bind(org_id)
        .bind(owner_id)
        .bind(&title)
        .bind(&body)
        .bind(payload.to_string())
        .execute(&state.pool)
        .await
        .map_err(AppError::Db)?;
        created += result.rows_affected() as i64;
    }

    // ── unowned overdue work goes to that org's managers ──
    // Scoping the aggregate and the fan-out per organization keeps a manager's
    // escalation from mixing another tenant's overdue counts and amounts.
    let orgs = sqlx::query("SELECT id FROM organizations WHERE is_active")
        .fetch_all(&state.pool)
        .await
        .map_err(AppError::Db)?;

    for org in &orgs {
        let org_id: Uuid = org
            .try_get("id")
            .map_err(|e| AppError::Internal(e.to_string()))?;

        let orphan = sqlx::query(
            "SELECT COUNT(*) AS n, SUM(d.charge_amount)::float8 AS amount, MIN(d.appeal_deadline) AS oldest \
               FROM denials d \
               JOIN claims c ON c.id = d.claim_id \
               LEFT JOIN appeals_queue aq ON aq.denial_id = d.id \
                    AND (aq.outcome_status IS NULL \
                         OR aq.outcome_status NOT IN ('approved','overruled','resolved','denied_again','cancelled')) \
              WHERE c.organization_id = $1 \
                AND d.appeal_deadline IS NOT NULL \
                AND d.appeal_deadline < CURRENT_DATE \
                AND d.status IN ('open', 'analyzed') \
                AND (aq.id IS NULL OR aq.assigned_user_id IS NULL)",
        )
        .bind(org_id)
        .fetch_one(&state.pool)
        .await
        .map_err(AppError::Db)?;

        let orphan_n: i64 = orphan.try_get("n").unwrap_or(0);
        if orphan_n > 0 {
            let oldest: Option<NaiveDate> = orphan.try_get("oldest").ok().flatten();
            let amount: f64 = orphan
                .try_get::<Option<f64>, _>("amount")
                .ok()
                .flatten()
                .unwrap_or(0.0);
            let managers = sqlx::query(
                "SELECT om.user_id AS id \
                       FROM organization_memberships om \
                      WHERE om.organization_id = $1 \
                        AND om.role = ANY($2::text[]) \
                        AND om.user_id IN (SELECT id FROM users WHERE is_active)",
            )
            .bind(org_id)
            .bind(MANAGER_UP.iter().map(|s| s.to_string()).collect::<Vec<_>>())
            .fetch_all(&state.pool)
            .await
            .map_err(AppError::Db)?;

            let title = format!("{orphan_n} overdue denial(s) with nobody assigned");
            let body = format!(
                "Oldest deadline {}. {} in denied charges is unowned.",
                oldest.map(|d| d.to_string()).unwrap_or_default(),
                money(amount),
            );
            let payload = serde_json::json!({
                "count": orphan_n,
                "oldest": oldest.map(|d| d.to_string()),
            });

            for m in &managers {
                let mid: Uuid = m
                    .try_get("id")
                    .map_err(|e| AppError::Internal(e.to_string()))?;
                let result = sqlx::query(
                    "INSERT INTO notifications (organization_id, user_id, kind, title, body, payload) \
                     VALUES ($1, $2, 'overdue_escalation', $3, $4, $5::jsonb) \
                     ON CONFLICT (organization_id, user_id, kind, for_date) DO NOTHING",
                )
                .bind(org_id)
                .bind(mid)
                .bind(&title)
                .bind(&body)
                .bind(payload.to_string())
                .execute(&state.pool)
                .await
                .map_err(AppError::Db)?;
                escalated += result.rows_affected() as i64;
            }
        }

        // ── overpayments past their refund deadline (FB-08) ──
        let overdue_refunds = sqlx::query(
            "SELECT COUNT(*) AS n, SUM(amount)::float8 AS amount, MIN(due_date) AS oldest \
               FROM overpayments \
              WHERE organization_id = $1 AND status = 'identified' AND due_date < CURRENT_DATE",
        )
        .bind(org_id)
        .fetch_one(&state.pool)
        .await
        .map_err(AppError::Db)?;
        let refund_n: i64 = overdue_refunds.try_get("n").unwrap_or(0);
        if refund_n > 0 {
            let oldest: Option<NaiveDate> = overdue_refunds.try_get("oldest").ok().flatten();
            let amount: f64 = overdue_refunds
                .try_get::<Option<f64>, _>("amount")
                .ok()
                .flatten()
                .unwrap_or(0.0);
            let title = format!("{refund_n} overpayment(s) past their refund deadline");
            let body = format!(
                "{} identified but not refunded, recouped or disputed; oldest due {}.",
                money(amount),
                oldest.map(|d| d.to_string()).unwrap_or_default(),
            );
            let payload = serde_json::json!({
                "count": refund_n,
                "amount": amount,
                "oldest": oldest.map(|d| d.to_string()),
            });
            let result = sqlx::query(
                "INSERT INTO notifications (organization_id, user_id, kind, title, body, payload) \
                 SELECT $1, om.user_id, 'overpayment_due', $3, $4, $5::jsonb \
                   FROM organization_memberships om \
                  WHERE om.organization_id = $1 AND om.role = ANY($2::text[]) \
                    AND om.user_id IN (SELECT id FROM users WHERE is_active) \
                 ON CONFLICT (organization_id, user_id, kind, for_date) DO NOTHING",
            )
            .bind(org_id)
            .bind(MANAGER_UP.iter().map(|s| s.to_string()).collect::<Vec<_>>())
            .bind(&title)
            .bind(&body)
            .bind(payload.to_string())
            .execute(&state.pool)
            .await
            .map_err(AppError::Db)?;
            escalated += result.rows_affected() as i64;
        }
    }

    denial_audit::record(
        &state.pool,
        "deadline_digests_generated",
        "system",
        None,
        None,
        &serde_json::json!({
            "digests": created,
            "escalations": escalated,
            "horizon_days": horizon,
        }),
        None,
        None,
        principal.organization_id.as_deref(),
    )
    .await;

    Ok(Json(serde_json::json!({
        "status": "ok",
        "digests_created": created,
        "escalations_created": escalated,
        "horizon_days": horizon,
    })))
}

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/", get(list_notifications))
        .route("/read-all", post(mark_all_read))
        .route("/generate-digests", post(generate_digests))
        .route("/{notification_id}/read", post(mark_read))
}
